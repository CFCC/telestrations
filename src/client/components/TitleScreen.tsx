import React, {ChangeEvent, Component} from 'react';
import {Button, CircularProgress, createStyles, TextField, Theme, WithStyles} from "@material-ui/core";
import {connectAndStyle} from "../util";
import {State} from "../redux/reducers";
import * as Creators from "../redux/actions";
import {ClientGameState} from "../../types";

const styles = (theme: Theme) => createStyles({
    app: {
        backgroundColor: '#FFC20E',
        display: 'flex',
        alignItems: 'center',
        flexDirection: 'column',
        height: '100vh',
        overflow: 'auto',
        padding: '1em',
        marginBottom: '1em'
    },
    progress: {
        margin: theme.spacing.unit * 5
    },
    img: {
        maxWidth: '80%'
    },
    input: {
        marginBottom: '1em'
    },
    header: {
        textAlign: 'center'
    }
});

const mapStateToProps = (state: State) => ({
    nickname: state.nickname,
    state: state.state,
    nicknameSubmitted: state.nicknameSubmitted
});

const mapDispatchToProps = {
    submitNickname: Creators.submitNickname,
    setNickname: (nickname: String) => Creators.setNickname(nickname)
};

type LoadingScreenProps = WithStyles<typeof styles> & ReturnType<typeof mapStateToProps> & typeof mapDispatchToProps;

class TitleScreen extends Component<LoadingScreenProps> {
    logo = <img src="/logo.png" alt="Telestrations logo" className={this.props.classes.img} />;

    submitNickname = () => [
        <h1 className={this.props.classes.header}>What is your name?</h1>,
        <TextField value={this.props.nickname}
                   variant="outlined"
                   className={this.props.classes.input}
                   onChange={(e: ChangeEvent<HTMLInputElement>) => this.props.setNickname(e.target.value)} />,
        <Button onClick={this.props.submitNickname} variant="contained" color="primary">Join Game</Button>
    ];

    waitForGameToStart = () => [
        <h1 className={this.props.classes.header}>Waiting for the game to start</h1>,
        <h3>Have your host start the game when everyone's joined!</h3>,
        <CircularProgress className={this.props.classes.progress} />
    ];

    gameAlreadyStarted = () => [
        <h1 className={this.props.classes.header}>This game's already started!</h1>,
        <h3>Wait for it to finish before joining.</h3>
    ];

    gameFinished = () => [
        <h1 className={this.props.classes.header}>The game is finished!</h1>,
        <h3>Please ask your host to see the results.</h3>
    ];

    getContent = () => {
        switch(this.props.state) {
            case ClientGameState.LOADING:
                return this.props.nicknameSubmitted ? this.waitForGameToStart() : this.submitNickname();
            case ClientGameState.FINISHED:
                return this.gameFinished();
            case ClientGameState.ALREADY_STARTED:
            default:
                return this.gameAlreadyStarted();
        }
    };

    render() {
        return <div className={this.props.classes.app}>
            {this.logo}
            {this.getContent()}
        </div>;
    }
}

export default connectAndStyle(TitleScreen, mapStateToProps, mapDispatchToProps, styles);