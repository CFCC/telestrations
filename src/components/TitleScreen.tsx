import React from "react";
import {CircularProgress, Typography} from "@material-ui/core";
import styled from "styled-components";

const Container = styled.div`
    display: flex;
    align-items: center;
    flex-direction: column;
`;

const Progress = styled(CircularProgress)`
    margin: 1rem;
    width: 3rem;
    height: 3rem;
`;

const Image = styled.img`
    max-width: 50%;
    margin: 1rem;
`;

const Header = styled(Typography)`
    text-align: center;
    font-size: 2rem;
    margin: 1rem;
    font-weight: bold;
`;

interface TitleScreenProps {
    title: string;
    subtitle?: string;
    loading?: boolean;
    children?: React.ReactNode;
}

export default function TitleScreen({title, subtitle, loading = false, children}: TitleScreenProps) {
    return (
        <Container>
            <Image src="/logo.png" alt="Telestrations logo" />
            <Header>{title}</Header>
            {subtitle && <Typography>{subtitle}</Typography>}
            {children}
            {loading && <Progress />}
        </Container>
    );
}
