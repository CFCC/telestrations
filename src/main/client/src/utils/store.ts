import { createGamekitSlice, sendEvent } from "@piticent123/gamekit-client";
import {
  configureStore,
  createAsyncThunk,
  createSlice,
  PayloadAction,
} from "@reduxjs/toolkit";
import {
  TypedUseSelectorHook,
  useSelector as useUntypedSelector,
} from "react-redux";

import { GameState, TelestrationsGame, TelestrationsPlayer } from "./types";
import _ from "lodash";
import { Settings } from "@piticent123/gamekit-client/lib/types";

const { reducer: gamekitReducer, actions: gamekitActions } =
  createGamekitSlice<Settings, TelestrationsPlayer, TelestrationsGame>();

interface AppState {
  gameState: GameState;
  activeContentId: string;
}
interface State {
  app: AppState;
  gamekit: ReturnType<typeof gamekitReducer>;
}
interface ThunkApi {
  state: State;
}

// This has to be at the top level to be throttled correctly
const _sendWritePageEvent = _.throttle((gameCode: string, content: string) => {
  sendEvent({
    route: `/app/games/${gameCode}/page/write`,
    data: content,
  });
}, 500);
export const setGuess = createAsyncThunk<void, string, ThunkApi>(
  "setGuess",
  async (content, { getState }) => {
    const {
      gamekit: { currentGame },
    } = getState();

    _sendWritePageEvent(currentGame.code, content);
  }
);

export const submitGuess = createAsyncThunk<void, string, ThunkApi>(
  "submitGuess",
  async (content, { getState }) => {
    const {
      gamekit: { currentGame },
    } = getState();

    sendEvent({
      route: `/app/games/${currentGame.code}/page/submit`,
      data: content,
    });
  }
);

export const startGame = createAsyncThunk<void, string, ThunkApi>(
  "startGame",
  async (__, { getState }) => {
    const {
      gamekit: { currentGame },
    } = getState();

    sendEvent({ route: `/games/${currentGame.code}/start` });
  }
);

export const {
  createGame,
  joinGame,
  rejoinGame,
  connectToServer,
  saveSettings,
  becomeAdmin,
} = gamekitActions;

const { actions, reducer: appReducer } = createSlice({
  name: "app",
  initialState: {
    gameState: GameState.LOGIN,
    activeContentId: "",
  } as AppState,
  reducers: {
    viewPlayerHistory: (state, action: PayloadAction<string>) => {
      state.gameState = GameState.PLAYER_HISTORY;
      state.activeContentId = action.payload;
    },
    viewNotepadHistory: (state, action: PayloadAction<string>) => {
      state.gameState = GameState.NOTEPAD_HISTORY;
      state.activeContentId = action.payload;
    },
    setGameState: (state, action: PayloadAction<GameState>) => {
      state.gameState = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(gamekitActions.rejoinGame.fulfilled, (state) => {
        state.gameState = GameState.IN_GAME;
      })
      .addCase(gamekitActions.connectToServer.fulfilled, (state) => {
        state.gameState = GameState.GAME_CODE;
      })
      .addCase(gamekitActions.joinGame.fulfilled, (state) => {
        state.gameState = GameState.WAITING_TO_START;
      })
      .addCase(submitGuess.fulfilled, (state) => {
        state.gameState = GameState.WAITING_FOR_CONTENT;
      });
  },
});

export const {
  viewNotepadHistory,
  viewPlayerHistory,
  setGameState
} = actions;

export const store = configureStore({
  reducer: { gamekit: gamekitReducer, app: appReducer },
});

export const useSelector: TypedUseSelectorHook<State> = useUntypedSelector;
