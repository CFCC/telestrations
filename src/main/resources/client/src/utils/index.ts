export function sleep(t: number): Promise<void> {
  return new Promise((res) => setTimeout(res, t));
}
