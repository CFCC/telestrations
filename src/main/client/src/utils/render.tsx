import React from "react";
import { render } from "@testing-library/react";
import { StylesProvider, ThemeProvider } from "@material-ui/core/styles";
import { Provider } from "react-redux";

import { GlobalStyles, theme } from "./theme";
import { store } from "./store";

interface WrapperProps {
  children: React.ReactNode;
}

const wrapper = ({ children }: WrapperProps) => (
  <React.Fragment>
    <StylesProvider injectFirst={true}>
      <ThemeProvider theme={theme}>
        <GlobalStyles />
        <Provider store={store}>{children}</Provider>
      </ThemeProvider>
    </StylesProvider>
  </React.Fragment>
);
const customRender = (ui: React.ReactElement) =>
  render(ui, { wrapper } as object);

export { customRender as render };
