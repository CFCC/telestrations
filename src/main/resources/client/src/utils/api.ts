import SockJS from "sockjs-client";
import { Client } from "@stomp/stompjs";
import _ from "lodash";
import { v4 as uuid } from "uuid";

import { store, actions, rejoinGame } from "./store";
import { Game, Settings } from "./types";

const client = new Client({
  webSocketFactory: () => new SockJS(`${location.origin}/websocket-server`),
});
client.onConnect = () => {
  client.subscribe("/topic/rejoin-game", ({ body }) => {
    store.dispatch(rejoinGame(body));
  });

  client.subscribe("/topic/games", ({ body }) => {
    const response = JSON.parse(body) as string[];
    store.dispatch(actions.handleGamesListMessage(response));
  });

  client.subscribe(`/user/topic/errors/client`, ({ body }) => {
    store.dispatch(actions.handleRequestException(body));
  });

  client.subscribe(`/user/topic/errors/server`, ({ body }) => {
    store.dispatch(actions.handleServerException(body));
  });

  client.subscribe(`/user/topic/successes`, ({ body }) => {
    store.dispatch(actions.handleSuccess(body));
  });
};

export const getGameCodes = (callback: (ids: string[]) => void) => {
  return firebase
    .firestore()
    .collection("games")
    .where("status", "==", "lobby")
    .onSnapshot(async function (snapshot) {
      const newGames = snapshot.docs.map((doc) => doc.id);
      callback(newGames);
    });
};
export const setGameCode = (gameCode: string, isClient: boolean = false) => {
  const game = firebase
    .firestore()
    .doc(`games/${gameCode}`) as DocumentReference<Partial<Game>>;

  game.onSnapshot(async (snapshot) => {
    const newGame = snapshot.data();
    if (!newGame) return;

    if (isClient) {
      if (newGame.status === "in progress") {
        store.dispatch(setGameState(GameState.IN_GAME));
      } else if (newGame.status === "finished") {
        store.dispatch(setGameState(GameState.FINISHED));
        const currentUser = firebase.auth().currentUser;
        if (currentUser?.isAnonymous) {
          await currentUser.delete();
        }
      }
    }

    store.dispatch(updateGame({ ...newGame, id: snapshot.id }));
  });
  game.collection("notepads").onSnapshot((snapshot) => {
    const notepads = _.mapValues(
      _.keyBy(snapshot.docs, "id"),
      (d) => d.data() as Notepad
    );
    store.dispatch(updateNotepads(notepads));
  });
  game.collection("players").onSnapshot(async (snapshot) => {
    const players = _.mapValues(
      _.keyBy(snapshot.docs, "id"),
      (d) => d.data() as Player
    );
    store.dispatch(updatePlayers(players));

    if (!isClient) {
      const {
        notepads,
        game: { status },
      } = store.getState().firebase;
      const gameIsOver =
        status === "in progress" &&
        Object.entries(players).every(
          ([pid, p]) => notepads[p.currentNotepad]?.ownerId === pid
        ) &&
        Object.values(notepads).every((n) => n.pages.length > 1);

      if (gameIsOver) await game.set({ status: "finished" }, { merge: true });
    } else {
      const { uid } = store.getState().client.user;
      const { currentNotepad, queue } = players[uid];

      if (!currentNotepad && queue?.length > 0) {
        const newQueue = [...queue];
        const newNotepad = newQueue.shift();

        await (firebase
          .firestore()
          .doc(`games/${gameCode}/players/${uid}`) as DocumentReference<
          Partial<Player>
        >).set(
          { queue: newQueue, currentNotepad: newNotepad },
          { merge: true }
        );
        store.dispatch(clientSlice.actions.setGameState(GameState.IN_GAME));
      }
    }
  });
};
export const startGame = async () => {
  const {
    firebase: {
      game: { id: gameCode },
      players,
    },
  } = store.getState();
  const playerIds = Object.keys(players);

  await Promise.all(
    playerIds.map(async (playerId, i) => {
      const notepadId = uuid();
      await (firebase
        .firestore()
        .doc(
          `games/${gameCode}/notepads/${notepadId}`
        ) as DocumentReference<Notepad>).set({ ownerId: playerId, pages: [] });
      await (firebase
        .firestore()
        .doc(`games/${gameCode}/players/${playerId}`) as DocumentReference<
        Partial<Player>
      >).set(
        {
          currentNotepad: notepadId,
          nextPlayer: playerIds[(i + 1) % playerIds.length],
          queue: [],
        },
        { merge: true }
      );
    })
  );
  await (firebase.firestore().doc(`games/${gameCode}`) as DocumentReference<
    Partial<Game>
  >).set({ status: "in progress" }, { merge: true });
};
export const setGuess = _.throttle(async (guess: string) => {
  const {
    client: { user },
    firebase: {
      game: { id: gameCode },
      players,
      notepads,
    },
  } = store.getState();
  const currentNotepadId = players[user.uid].currentNotepad;
  const currentNotepad = notepads[currentNotepadId];
  if (!currentNotepad) return;

  const pages = _.cloneDeep(currentNotepad.pages);
  const needsNewPage = _.last(pages)?.author !== user.uid;
  if (needsNewPage) {
    pages.push({
      author: user.uid,
      lastUpdated: new Date().getTime(),
      content: "",
    });
  }

  if (pages.length % 2 === 1) {
    pages[pages.length - 1].content = guess;
    pages[pages.length - 1].lastUpdated = new Date().getTime();
  } else {
    let fileName = _.last(pages)?.content;
    if (!fileName) {
      fileName = `${uuid()}.png`;
      pages[pages.length - 1].content = fileName;
    }

    pages[pages.length - 1].lastUpdated = new Date().getTime();
    firebase.storage().ref().child(fileName).putString(guess, "data_url");
  }

  await (firebase
    .firestore()
    .doc(`games/${gameCode}/notepads/${currentNotepadId}`) as DocumentReference<
    Partial<Notepad>
  >).set({ pages }, { merge: true });
}, 500);
export const submitGuess = async () => {
  const {
    client: { user },
    firebase: {
      game: { id: gameCode },
      players,
    },
  } = store.getState();
  const { currentNotepad, nextPlayer } = players[user.uid];
  const { queue: nextPlayerQueue } = players[nextPlayer];

  await (firebase
    .firestore()
    .doc(`games/${gameCode}/players/${nextPlayer}`) as DocumentReference<
    Partial<Player>
  >).set({ queue: [...nextPlayerQueue, currentNotepad] }, { merge: true });
  await (firebase
    .firestore()
    .doc(`games/${gameCode}/players/${user.uid}`) as DocumentReference<
    Partial<Player>
  >).set({ currentNotepad: "" }, { merge: true });
};

export function updateSettings(code: string, settings: Settings) {
  if (!code) return;

  client.publish({
    destination: `/app/games/${code}/update`,
    body: JSON.stringify(settings),
  });
}

export function connectToServer(uuid: string) {
  client.connectHeaders = { uuid };
  client.activate();
}

export function createGame(code: string) {
  client.publish({ destination: `/app/games/${code}/create` });
}

export async function joinGame(
  code: string,
  settings: Settings,
  rejoining: boolean = false
) {
  if (!rejoining) {
    client.publish({
      destination: `/app/games/${code}/join`,
      body: JSON.stringify(settings),
    });
  }

  // A race condition happens where if we dont wait 1ms, the subscription may happen first
  // giving an error that the game doesn't exist
  await new Promise((resolve) => setTimeout(resolve, 1));

  client.subscribe(`/topic/games/${code}`, ({ body }) => {
    const response = JSON.parse(body) as Game;
    store.dispatch(actions.handleGameUpdate(response));
  });
}
