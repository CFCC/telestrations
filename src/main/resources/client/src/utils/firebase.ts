import _ from "lodash";
import { v4 as uuid } from "uuid";
import { animals, colors, uniqueNamesGenerator } from "unique-names-generator";

import { clientSlice, firebaseSlice, GameState, store } from "./store";

const {
  actions: { updatePlayers, updateNotepads, updateGame },
} = firebaseSlice;
const {
  actions: { setGameState, setUser },
} = clientSlice;

firebase.auth().onAuthStateChanged(function (user: firebase.User | null) {
  if (!user) return;

  let { displayName, uid } = user;
  if (!user.displayName) {
    displayName = _.startCase(
      uniqueNamesGenerator({
        dictionaries: [colors, animals],
        length: 2,
        separator: "-",
      })
    );
    user.updateProfile({ displayName });
  }

  store.dispatch(setUser({ uid, displayName }));
});

export const getImageURL = async (image: string): Promise<string> => {
  return (await firebase.storage().ref(image).getDownloadURL()) as string;
};
export const getAllGameCodes = async () => {
  return (await firebase.firestore().collection("games").get()).docs;
};
export const getGameCodes = (callback: (ids: string[]) => void) => {
  return firebase
    .firestore()
    .collection("games")
    .where("status", "==", "lobby")
    .onSnapshot(async function (snapshot) {
      const newGames = snapshot.docs.map((doc) => doc.id);
      callback(newGames);
    });
};
export const setGameCode = (gameCode: string, isClient: boolean = false) => {
  const game = firebase
    .firestore()
    .doc(`games/${gameCode}`) as DocumentReference<Partial<Game>>;

  game.onSnapshot(async (snapshot) => {
    const newGame = snapshot.data();
    if (!newGame) return;

    if (isClient) {
      if (newGame.status === "in progress") {
        store.dispatch(setGameState(GameState.IN_GAME));
      } else if (newGame.status === "finished") {
        store.dispatch(setGameState(GameState.FINISHED));
        const currentUser = firebase.auth().currentUser;
        if (currentUser?.isAnonymous) {
          await currentUser.delete();
        }
      }
    }

    store.dispatch(updateGame({ ...newGame, id: snapshot.id }));
  });
  game.collection("notepads").onSnapshot((snapshot) => {
    const notepads = _.mapValues(
      _.keyBy(snapshot.docs, "id"),
      (d) => d.data() as Notepad
    );
    store.dispatch(updateNotepads(notepads));
  });
  game.collection("players").onSnapshot(async (snapshot) => {
    const players = _.mapValues(
      _.keyBy(snapshot.docs, "id"),
      (d) => d.data() as Player
    );
    store.dispatch(updatePlayers(players));

    if (!isClient) {
      const {
        notepads,
        game: { status },
      } = store.getState().firebase;
      const gameIsOver =
        status === "in progress" &&
        Object.entries(players).every(
          ([pid, p]) => notepads[p.currentNotepad]?.ownerId === pid
        ) &&
        Object.values(notepads).every((n) => n.pages.length > 1);

      if (gameIsOver) await game.set({ status: "finished" }, { merge: true });
    } else {
      const { uid } = store.getState().client.user;
      const { currentNotepad, queue } = players[uid];

      if (!currentNotepad && queue?.length > 0) {
        const newQueue = [...queue];
        const newNotepad = newQueue.shift();

        await (firebase
          .firestore()
          .doc(`games/${gameCode}/players/${uid}`) as DocumentReference<
          Partial<Player>
        >).set(
          { queue: newQueue, currentNotepad: newNotepad },
          { merge: true }
        );
        store.dispatch(clientSlice.actions.setGameState(GameState.IN_GAME));
      }
    }
  });
};
export const createGame = async (gameCode: string) => {
  const {
    client: { user },
  } = store.getState();

  await (firebase
    .firestore()
    .doc(`games/${gameCode}`) as DocumentReference<Game>).set({
    created: new Date().getTime(),
    status: "lobby",
    serverId: user.uid,
  } as Game);
};
export const joinGame = async (gameCode: string) => {
  const {
    client: { user },
  } = store.getState();

  await (firebase
    .firestore()
    .doc(`games/${gameCode}/players/${user.uid}`) as DocumentReference<
    Partial<Player>
  >).set({ name: user.displayName as NonNullable<string> });
};
export const startGame = async () => {
  const {
    firebase: {
      game: { id: gameCode },
      players,
    },
  } = store.getState();
  const playerIds = Object.keys(players);

  await Promise.all(
    playerIds.map(async (playerId, i) => {
      const notepadId = uuid();
      await (firebase
        .firestore()
        .doc(
          `games/${gameCode}/notepads/${notepadId}`
        ) as DocumentReference<Notepad>).set({ ownerId: playerId, pages: [] });
      await (firebase
        .firestore()
        .doc(`games/${gameCode}/players/${playerId}`) as DocumentReference<
        Partial<Player>
      >).set(
        {
          currentNotepad: notepadId,
          nextPlayer: playerIds[(i + 1) % playerIds.length],
          queue: [],
        },
        { merge: true }
      );
    })
  );
  await (firebase.firestore().doc(`games/${gameCode}`) as DocumentReference<
    Partial<Game>
  >).set({ status: "in progress" }, { merge: true });
};
const updateGuessInFirebase = _.throttle(async (guess: string) => {
  const {
    client: { user },
    firebase: {
      game: { id: gameCode },
      players,
      notepads,
    },
  } = store.getState();
  const currentNotepadId = players[user.uid].currentNotepad;
  const currentNotepad = notepads[currentNotepadId];
  if (!currentNotepad) return;

  const pages = _.cloneDeep(currentNotepad.pages);
  const needsNewPage = _.last(pages)?.author !== user.uid;
  if (needsNewPage) {
    pages.push({
      author: user.uid,
      lastUpdated: new Date().getTime(),
      content: "",
    });
  }

  if (pages.length % 2 === 1) {
    pages[pages.length - 1].content = guess;
    pages[pages.length - 1].lastUpdated = new Date().getTime();
  } else {
    let fileName = _.last(pages)?.content;
    if (!fileName) {
      fileName = `${uuid()}.png`;
      pages[pages.length - 1].content = fileName;
    }

    pages[pages.length - 1].lastUpdated = new Date().getTime();
    firebase.storage().ref().child(fileName).putString(guess, "data_url");
  }

  await (firebase
    .firestore()
    .doc(`games/${gameCode}/notepads/${currentNotepadId}`) as DocumentReference<
    Partial<Notepad>
  >).set({ pages }, { merge: true });
}, 500);
export const setGuess = (guess: string) => updateGuessInFirebase(guess);
export const submitGuess = async () => {
  const {
    client: { user },
    firebase: {
      game: { id: gameCode },
      players,
    },
  } = store.getState();
  const { currentNotepad, nextPlayer } = players[user.uid];
  const { queue: nextPlayerQueue } = players[nextPlayer];

  await (firebase
    .firestore()
    .doc(`games/${gameCode}/players/${nextPlayer}`) as DocumentReference<
    Partial<Player>
  >).set({ queue: [...nextPlayerQueue, currentNotepad] }, { merge: true });
  await (firebase
    .firestore()
    .doc(`games/${gameCode}/players/${user.uid}`) as DocumentReference<
    Partial<Player>
  >).set({ currentNotepad: "" }, { merge: true });
};
