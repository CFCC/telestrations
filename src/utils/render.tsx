import React from "react";
import {render} from "@testing-library/react";
import {StylesProvider, ThemeProvider} from "@material-ui/core/styles";
import {Provider} from "react-redux";

import {GlobalStyles, theme} from "./theme";
import {store} from "../store/client";

interface WrapperProps {
    children: React.ReactNode;
}

const wrapper = (store: any) => ({children}: WrapperProps) => (
    <React.Fragment>
        <StylesProvider injectFirst={true}>
            <ThemeProvider theme={theme}>
                <GlobalStyles />
                <Provider store={store}>
                    {children}
                </Provider>
            </ThemeProvider>
        </StylesProvider>
    </React.Fragment>
);
const customRender = (ui: React.ReactElement, options: any) => render(ui, {wrapper, ...options});

export {customRender as render};
