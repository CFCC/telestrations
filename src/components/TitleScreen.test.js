import {render} from "../utils/render";
import {Standard} from "./TitleScreen.stories";

it('Renders', () => {
    const {asFragment} = render(Standard());
    expect(asFragment()).toMatchSnapshot();
});
import React from "react";

