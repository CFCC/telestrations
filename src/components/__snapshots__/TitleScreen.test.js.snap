// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm kwsLvR"
  >
    <img
      alt="Telestrations"
      class="sc-AxiKw iXGiQC"
      src="/logo.png"
    />
    <p
      class="MuiTypography-root sc-AxhCb bDZFln MuiTypography-body1"
    >
      Hello, world!
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      How are you?
    </p>
  </div>
</DocumentFragment>
`;
